### PR TITLE
Refactor domain queue and add domain dlq size metrics

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Refactor domain queue and add domain DLQ size metrics

**Why?**
Current domain DLQ size calculates from two metrics, which is not reliable. As the queue size should be small. We can do count on this queue.

**How did you test it?**
integration tests

**Potential risks**
In the worst case, it failed to emit DLQ size metric

### CODE REVIEW

File: service/worker/service.go [hunk 1]
```diff
@@ -197,6 +197,7 @@ func (s *Service) Start() {
 	logger.Info("worker starting", tag.ComponentWorker)
 
 	s.Resource.Start()
+	s.Resource.GetDomainReplicationQueue().Start()
```
Comment (yycptt) [inline service/worker/service.go:200]: Is worker service the only service need DomainReplicationQueue? If so then probably we should consider move domain replication queue out of resource.

Comment (yux0) [reply service/worker/service.go:200]: Both frontend and worker will use it

File: common/persistence/nosql/nosqlplugin/interfaces.go [hunk 2]
```diff
@@ -103,6 +103,8 @@ type (
 		UpdateQueueMetadataCas(ctx context.Context, row QueueMetadataRow) error
 		// Read a QueueMetadata
 		SelectQueueMetadata(ctx context.Context, queueType persistence.QueueType) (*QueueMetadataRow, error)
+		// SelectQueueSize return the queue size
+		SelectQueueSize(ctx context.Context, queueType persistence.QueueType) (int64, error)
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/persistence/nosql/nosqlplugin/interfaces.go]: updated

File: common/persistence/cassandra/cassandraQueue.go [hunk 3]
```diff
@@ -301,6 +301,25 @@ func (q *nosqlQueue) GetDLQAckLevels(
...
+	if err != nil {
+		switch {
+		case q.db.IsThrottlingError(err):
+			return size, &shared.ServiceBusyError{Message: "GetDLQSize operation failed with transient error."}
+		default:
+			return size, &shared.InternalServiceError{
+				Message: fmt.Sprintf("GetDLQSize operation failed. Error %v", err),
+			}
+
+		}
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/persistence/cassandra/cassandraQueue.go]: Didn't know about this. Updated

File: common/domain/replication_queue.go [hunk 4]
```diff
@@ -343,3 +326,31 @@ func (q *domainReplicationQueueImpl) purgeProcessor() {
...
+				q.metricsClient.Scope(
+					metrics.DomainReplicationQueueScope,
+				).IncCounter(
+					metrics.DomainReplicationQueueSizeErrorCount,
+				)
+			}
+			q.metricsClient.Scope(
+				metrics.DomainReplicationQueueScope,
+			).AddCounter(
+				metrics.DomainReplicationQueueSizeCounter,
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/domain/replication_queue.go]: make sense. Updated

File: common/domain/replication_queue.go [hunk 5]
```diff
@@ -36,24 +36,25 @@ import (
...
 	localDomainReplicationCluster = "domainReplication"
 )
 
-var _ DomainReplicationQueue = (*domainReplicationQueueImpl)(nil)
+var _ ReplicationQueue = (*replicationQueueImpl)(nil)
 
-// NewDomainReplicationQueue creates a new DomainReplicationQueue instance
-func NewDomainReplicationQueue(
-	queue QueueManager,
+// NewReplicationQueue creates a new DomainReplicationQueue instance
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/domain/replication_queue.go]: updated

**Review by yycptt — APPROVED** [review]

**Pull request merged**